crates/driver/src/domain/competition/quote.rs

```diff
@@ -0,0 +1,182 @@
+    pub amount: Amount,
+    pub valid_to: u32,
+    /// See [`crate::domain::competition::order::Partial`].
+    pub partial: bool,
+    pub quality: Quality,
+    pub gas_price: eth::EffectiveGasPrice,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Amount {
+    Sell(eth::U256),
+    Buy(eth::U256),
+}
+
+impl Order {
+    /// The asset being bought. Zero if this is a sell.
+    pub fn buy(&self) -> eth::Asset {
+        match self.amount {
+            Amount::Sell(..) => eth::Asset {
+                amount: Default::default(),
+                token: self.buy_token,
+            },
+            Amount::Buy(amount) => eth::Asset {
+                amount,
+                token: self.buy_token,
+            },
+        }
+    }
+
+    /// The asset being sold. Zero if this is a buy.
+    pub fn sell(&self) -> eth::Asset {
+        match self.amount {
+            Amount::Sell(amount) => eth::Asset {
+                amount,
+                token: self.sell_token,
+            },
+            Amount::Buy(..) => eth::Asset {
+                amount: Default::default(),
+                token: self.sell_token,
+            },
+        }
+    }
+
+    pub fn side(&self) -> competition::order::Side {
+        match self.amount {
+            Amount::Sell(..) => competition::order::Side::Sell,
+            Amount::Buy(..) => competition::order::Side::Buy,
+        }
+    }
+}
+
+/// Quality of the quote to be generated. This value determines the time
+/// allocated for the solver to generate the solution.
+#[derive(Debug, Clone, Copy)]
+pub enum Quality {
+    Fast,
+    Optimal,
+}
+
+impl Order {
+    /// Generate a quote for this order. This calls `/solve` on the solver with
+    /// a "fake" auction which contains a single order, and then determines
+    /// the quote for the order based on the solution that the solver
+    /// returns.
+    pub async fn quote(&self, solver: &Solver, config: &Config) -> Result<Quote, Error> {
+        let solution = solver
+            .solve(&self.fake_auction(), self.timeout(config))
+            .await?;
+        let fulfillment = solution
+            .trades
+            .into_iter()
+            .filter_map(|trade| match trade {
+                competition::solution::Trade::Fulfillment(fulfillment) => Some(fulfillment),
+                competition::solution::Trade::Jit(..) => None,
+            })
+            .exactly_one()
+            .map_err(|_| Error::QuotingFailed)?;
+        Ok(Quote::new(fulfillment, solution.interactions))
+    }
+
+    fn fake_auction(&self) -> competition::Auction {
+        competition::Auction {
+            id: None,
+            tokens: Default::default(),
+            orders: vec![competition::Order {
+                uid: Default::default(),
+                receiver: None,
+                valid_to: util::Timestamp::MAX,
+                sell: self.sell(),
+                buy: self.buy(),
+                side: self.side(),
+                fee: Default::default(),
```

Comment:
The [current implementation](https://github.com/cowprotocol/services/blob/main/crates/shared/src/price_estimation/http.rs#L103-L110) adds a fee to the fake order it sends to the solver for quoting. (I didn't notice that we actually already do `gas_used * gas_price` for that.
But even if the fee here is supposed to be in the `sell_token` I think we still need to have that instead of 0.

Reply:
So, I think that was there to ensure economic viability of the order (i.e. the order needs to pay enough fee to be worth settling). I would call on @harisang for some guidance here, as I'm not entirely sure myself.

If my assumption is correct, I _think_ a more appropriate value for this would just be `sell_amount` as:
- If it is too high, no harm done - it just makes the order attractive to trade for the solver, so it should be in the solution and we get our price quote.
- If it is too low, then this means that the full order sell amount amount is not enough to pay for its execution, so we are quoting for too small an amount anyway.

Reply:
I wonder if we can bribe the solver with a COW `reward` instead of a `feeAmount`...

Reply:
> (i.e. the order needs to pay enough fee to be worth settling)

The current http solver API also has the flag `mandatory` on the order which sounds like it's supposed to achieve exactly that. That could indicate that there might be another reason to include some `fee` with the quote request.

Reply:
Issue with using this large values is that they can break token invariants that solvers may check for.

Actually, now that I think about it more, for tokens like COW, Quasimodo was producing price estimates, but `mean_gas_used_per_order * gas_price COW` is **WAY** less than the actual needed fee. This makes me think that it _might_ not actually be needed. Either way, I pinged @harisang and he should have a more definitive answer.

Reply:
Just to clarify a few things:

1. There is an external price vector, and the everything is summed up together in the obj value according to this external price vector. Is that correct?
2. For a sell order, is the minimum buy amount that a solver sees in the json it receives set to zero?
3. How is the (gas) cost term in the objective currently computed? Or is it not needed for ranking the different estimations?

Reply:
> There is an external price vector, and the everything is summed up together in the obj value according to this external price vector. Is that correct?

Not for price quotes.

> For a sell order, is the minimum buy amount that a solver sees in the json it receives set to zero?

It is set to 1.

> How is the (gas) cost term in the objective currently computed? Or is it not needed for ranking the different estimations?

It is not needed.

> How is the actual fee computed after we receive the price estimates?

`gas_estimate * gas_price * sell_token_price`.

Reply:
In any case, having a predetermined fee term in the obj only shifts the obj by a constant. So, it doesn't change anything at all.

Specifically for Quasimodo, if the mandatory flag is set to true, then it will pick on the order necessarily and attempt to fill it so as to maximize the objective. So in that case (flag is set to true), there is no need for a fee as well, since again, it just shifts the obj by a bit but does not change the optimization process and the final computed amounts.

Reply:
So, if there are not external prices, then what is the sell_token_price for example that we use to compute the final fee?

Reply:
It is computed via a `/solve` request like this one. Basically, a quote is two `/solve` requests:
1. Solve with a fake order with the user's actual traded amounts
2. Solve with a fake "buy 1 ETH" order selling the order's sell token

Reply:
Ok, let's leave the `fee_amount` as 0 for now - solvers aren't expected to need to optimize for economic viability for market orders. If this becomes a problem, we can re-evaluate.